bee-pow/src/score.rs

```diff
@@ -3,43 +3,81 @@
 
 //! Contains utilities to score Proof of Work.
 
-use bee_crypto::ternary::{
-    sponge::{CurlP81, Sponge},
-    HASH_LENGTH,
-};
 use bee_ternary::{b1t6, Btrit, T1B1Buf, TritBuf, Trits, T1B1};
 
-use crypto::hashes::{blake2b::Blake2b256, Digest};
+use crypto::hashes::{
+    blake2b::Blake2b256,
+    ternary::{curl_p::CurlP, HASH_LENGTH},
+    Digest,
+};
 
-/// Computes the Proof of Work score of given bytes.
-/// Panic: expects at least 8 bytes.
-pub fn compute_pow_score(bytes: &[u8]) -> f64 {
-    debug_assert!(bytes.len() >= std::mem::size_of::<u8>());
+/// Encapsulates the different steps that are used for scoring Proof of Work.
+pub struct PoWScorer {
+    blake2b: Blake2b256,
+    pow_input: TritBuf<T1B1Buf>,
+    curl: CurlP,
+}
+
+impl PoWScorer {
+    /// Creates an new `PoWScorer` that holds the required hash functions as internal state.
+    pub fn new() -> Self {
+        Self {
+            blake2b: Blake2b256::new(),
+            pow_input: TritBuf::<T1B1Buf>::with_capacity(HASH_LENGTH),
+            curl: CurlP::new(),
+        }
+    }
+
+    /// Returns the Proof of Work hash of given bytes.
+    /// Panic: expects at least 8 bytes.
+    pub fn hash(&mut self, bytes: &[u8]) -> TritBuf<T1B1Buf> {
+        debug_assert!(bytes.len() >= std::mem::size_of::<u8>());
+
+        // Compute Blake2b-256 hash of the message, excluding the nonce.
+        let length = bytes.len() - std::mem::size_of::<u64>();
+        let (head, tail) = bytes.split_at(length);
+        self.blake2b.update(head);
+        let pow_digest = self.blake2b.finalize_reset();
 
-    pow_score(&pow_hash(bytes), bytes.len())
+        // Encode message as trits
+        self.pow_input.clear();
+        b1t6::encode::<T1B1Buf>(&pow_digest)
+            .iter()
+            .for_each(|t| self.pow_input.push(t));
+        b1t6::encode::<T1B1Buf>(tail)
+            .iter()
+            .for_each(|t| self.pow_input.push(t));
+
+        // Pad to 243 trits
+        self.pow_input.push(Btrit::Zero);
+        self.pow_input.push(Btrit::Zero);
+        self.pow_input.push(Btrit::Zero);
```

Comment:
I think you could also use the `append` API here on `TritBuf`:

`self.pow_input.append(b1t6::encode::<T1B1buf>(&pow_digest).as_slice());`